### PR TITLE
fix: make _arm_auto_merge actually enqueue ready PRs (event-driven path)

### PR DESCRIPTION
**Pull request by loganfinney27**
**Agent:** Claude
**Date:** 2026-06-20
**Branch:** `claude/wire-review-thread-recompute`

> **Note:** this PR was reopened and **repurposed**. Its original change (a `pull_request_review_thread` workflow trigger) was invalid — that event is a GitHub *webhook* but not a supported Actions `on:` trigger (actionlint hard error; confirmed on the thread by CodeRabbit). That change is **fully reverted** here; `review-feedback-loop.yml` is now identical to `main`. What remains is the real fix the #508 field note identified.

---

## Changes Made

- **`.github/scripts/review_feedback_loop.py`** — fix `_arm_auto_merge` so it actually enqueues on a merge-queue repo:
  - New `_auto_merge_state(owner, repo, pr)` reads `autoMergeRequest.enabledAt` + `mergeQueueEntry` via GraphQL (`_graphql`), fail-open to `(False, False)`.
  - `_arm_auto_merge(owner, repo, pr)`:
    - **off** → plain `--auto` (as before);
    - **armed but not queued** (the #508 stuck state) → `disable → re-enable` toggle to re-fire the ready-transition;
    - **already queued** → no-op (disabling would evict it and restart its run).
  - Signature gains `owner, repo`, threaded through all three callers (`_maybe_arm_auto_merge`, `_build_reconciliation_report`, `enable_auto_merge`).
- **`tests/test_review_feedback_loop.py`** — 3 new tests lock the off / armed-not-queued / queued paths; signature-dependent assertions updated.

## Why

`_arm_auto_merge` was a single idempotent `gh pr merge --auto`. On a merge-queue repo that only calls `enablePullRequestAutoMerge`, and re-enabling on an already-armed PR is a no-op (`! The merge strategy for main is set by the merge queue`, exit 0) that never re-fires the transition which enqueues the PR. So a PR armed while blocked (most of the agent backlog) never enters the queue when it goes green — the **same defect #591 fixed in the batch shell loop**, but here on the **event-driven path** that `auto-merge-engage.yml`, `review-response.yml`, and `sync-pr` all use. This is the load-bearing half of the field note on #588.

## Verification

- `python3 -m unittest tests.test_review_feedback_loop` → **95 tests green**.
- `python3 -m py_compile` clean.
- The GraphQL `mergeQueueEntry` field is the one uncertainty I can't exercise offline; if it were ever invalid the query raises and `_auto_merge_state` fail-opens to `(False, False)` → plain enable (old behavior, no eviction risk). Worth a live `workflow_dispatch` confirmation on a real already-armed PR.

## Related Work

- #588 field note — the verified diagnosis this implements.
- #591 — the same toggle + honest counting in the batch `gh` loop.
- #508 — landed via the disable→enable toggle this generalizes.

## Blockers

- This branch touches `.github/scripts/` (a protected/governance path) and was originally the invalid-trigger PR, so it needs your review + manual handling to land.
- Deliberately **not** included: enqueue-verification (polling `merge_group` on the head sha) — separable robustness add-on, can follow.

---

**Checklist:**
- [x] Tests pass — 95 green, incl. 3 new toggle-path tests.
- [x] No secrets in diff
- [x] Documentation updated IF needed — rationale documented inline.
- [ ] Reviewer assigned

---

**Risk Level:**
- [x] HIGH: changes the live event-driven auto-merge arm path. Recommend a `workflow_dispatch` sanity check before relying on it.

---

**Labels to apply:**
- `agent:claude-code`

---

**Ready for Logan to review and merge.**

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Chores**
  * Enhanced auto-merge state tracking by checking current enablement and queue status before arming.
  * Updated re-enable behavior to skip work when already queued, and to reliably recover “stuck” PRs via a checked disable-then-rearm flow.
  * Improved reconciliation and reporting so already-armed (but not queued) PRs are handled correctly, and failures in the disable step are surfaced.
* **Tests**
  * Updated unit and orchestration tests for the new arming inputs and added coverage for queue/edge cases and error handling.

### CODE REVIEW

File: .github/scripts/review_feedback_loop.py [hunk 1]
```diff
@@ -157,7 +157,62 @@ def _run(cmd: list[str], check: bool = True) -> subprocess.CompletedProcess[str]
...
+                  mergeQueueEntry { id }
+                }
+              }
+            }
+            """,
+            owner=owner,
+            name=repo,
+            number=pr_number,
+        )
+    except RuntimeError:
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply .github/scripts/review_feedback_loop.py]: Addressed in `7984168b` — `_auto_merge_state` now catches `(RuntimeError, ValueError)`, so a malformed / JSON-decode payload from `_graphql` fails open to `(False, False)` (plain enable), matching the docstring's contract. Good catch.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/scripts/review_feedback_loop.py [hunk 2]
```diff
@@ -157,7 +157,62 @@ def _run(cmd: list[str], check: bool = True) -> subprocess.CompletedProcess[str]
...
+    # enablePullRequestAutoMerge. Re-enabling it on a PR that ALREADY has
+    # auto-merge is an idempotent no-op ("! The merge strategy for main is set by
+    # the merge queue", exit 0) that never re-fires the ready-transition which
+    # actually enqueues the PR — so a PR armed while it was blocked never enters
+    # the queue when it later goes green (verified on #508). If auto-merge is on
+    # but the PR is NOT yet queued, toggle off->on to re-fire the transition. If
+    # it is already queued, leave it alone — disabling would evict it and restart
+    # its queue run.
+    enabled, queued = _auto_merge_state(owner, repo, pr_number)
+    if enabled and queued:
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply .github/scripts/review_feedback_loop.py:209]: Deferring this one — valid, but minor in practice and the cleaner fix is a broader refactor:

- The extra `_auto_merge_state` call only fires on the **arm path** (eligible + labeled + not-blocked + non-protected), not for every PR.
- In the event-driven path (`sync_pr`, `review-submitted`) it's a single call per event — negligible.
- The N case is `reconcile-open-prs`, which runs only in `agent-review-gate.yml` — currently `workflow_dispatch`-only and disabled (its header documents that it runs the retired risk-tier schema and must not auto-fire).

Threading `autoMergeRequest`/`mergeQueueEntry` through the shared `_fetch_pr` query (consumed across many flows) is a wider change I'd rather not bundle into this live-path correctness fix. Tracking it as a follow-up optimization to revisit if/when reconcile is re-enabled at scale.

---
_Generated by [Claude Code](https://claude.ai/code)_
